README.md
# angular-eha.image

[![Build Status](https://travis-ci.org/eHealthAfrica/angular-eha.image.svg)](https://travis-ci.org/eHealthAfrica/angular-eha.image) ![Dependecy Status](https://david-dm.org/eHealthAfrica/angular-eha.image.svg) ![Dev Dependecy Status](https://david-dm.org/eHealthAfrica/angular-eha.image/dev-status.svg)

## Usage

## Installation

Install with npm:

    npm install --save angular-eha.image

Or alternatively bower:

    bower install --save angular-eha.image

The distribution bundle comes in 2 flavours, one with templates, one without - as well as providing both minified and unminfied versions of both, so take your pick:

- *dist/angular-eha.image.directive.js*
- *dist/angular-eha.image.directive.min.js*
- *dist/angular-eha-back-button.template.js* *(default)*
- *dist/angular-eha-back-button.template.min.js*

Then simply add `eha.image.directive` and `eha.image.directive.templates` (if you want templates!) as a dependencies somewhere in your project that makes sense and you're good to go.

#### A note on wiredep

If you're using wiredep `dist/angular-eha.image.directive.template.js` will be injected by default. If you don't want that to happen you'll like want to employ something along the following lines in your `Gruntfile`:

```javascript
wiredep: {
 ...
  options: {
    exclude: [
      'bower_components/angular-eha.image/dist/angular-eha.image.directive.templates.js'
    ]
  }
  ...
}
```

Then you're free to include whichever bundle you prefer in what ever manner you prefer.

### Example

```html
<html ng-app="backButtonExample">
  <head>
    <title>Back Button Example</title>
    <script src="bower_components/angular/angular.js"></script>
    <script src="bower_components/angular/angular-eha.image.directive.js"></script>
    <script>
      angular.module('ehaImageExample', ['eha.image.directive', 'eha.image.directive.templates'])
        .controller('DemoCtrl1', function($scope) {
           $scope.image = {
             title: 'Test Image',
             mimeType: 'image/jpeg',
             type: 'jpeg',
             data: '/9j/4AAQSkZJRgABAQEASABIAAD/4QBARXhpZgAATU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAAqACAAQAAAABAAABGqADAAQAAAABAAABGgAAAAD/7QA4UGhvdG9zaG9wIDMuMAA4QklNBAQAAAAAAAA4QklNBCUAAAAAABDUHYzZjwCyBOmACZjs+EJ+/8AAEQgBGgEaAwEiAAIRAQMRAf/EAB8AAAEFAQEBAQEBAAAAAAAAAAABAgMEBQYHCAkKC//EALUQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+v/EAB8BAAMBAQEBAQEBAQEAAAAAAAABAgMEBQYHCAkKC//EALURAAIBAgQEAwQHBQQEAAECdwABAgMRBAUhMQYSQVEHYXETIjKBCBRCkaGxwQkjM1LwFWJy0QoWJDThJfEXGBkaJicoKSo1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoKDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uLj5OXm5+jp6vLz9PX29/j5+v/bAEMAAQEBAQEBAgEBAgMCAgIDBAMDAwMEBQQEBAQEBQYFBQUFBQUGBgYGBgYGBgcHBwcHBwgICAgICQkJCQkJCQkJCf/bAEMBAQEBAgICBAICBAkGBQYJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCf/dAAQAEv/aAAwDAQACEQMRAD8A++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Q++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//R++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//S++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//T++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//U++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//V++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//W++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//X++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Q++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//R++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//S++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//T++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//U++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//V++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//W++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//X++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Q++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Z'
           };
        });
    </script>
  </head>
  <body>
    <eha-img image="image">
  </body>
</html>
```

### Example 2
```html
<html ng-app="backButtonExample">
  <head>
    <title>Back Button Example</title>
    <script src="bower_components/angular/angular.js"></script>
    <script src="bower_components/angular/angular-eha.image.directive.js"></script>
    <script>
      angular.module('ehaImageExample', ['eha.image.directive', 'eha.image.directive.templates']);
    </script>
  </head>
  <body>
    <eha-img src="http://placekitten.com/200/300">
  </body>
</html>
```

## Contributing

### Prerequisites

- Firefox (for running test suite)
- node (0.12.0)
- bower (1.3.12)
- grunt-cli (0.1.7)
- grunt (0.4.5)


### Installation

```bash
# Fork the upstream repo on github and pull down your fork
git clone git@github.com:yourusername/angular-eha.image.directive.git
# change into project folder
cd angular-eha.image.directive
# Add the upstream as a remote
git remote add upstream  git@github.com:eHealthAfrica/angular-eha.image.directive.git
# Install the dev dependencies
npm install
```

### Docs

Code should be documented following the guidelines set out by [jsdoc](http://usejsdoc.org/) and [ngdoc](https://github.com/angular/angular.js/wiki/Writing-AngularJS-Documentation). We can then leverage [Dgeni](http://github.com/angular/dgeni) or something simlary to generate documentation in any format we like.

### Test Suite

The test suite is configured to run in Firefox and is powered by:

- Karma
- Mocha
- Chai (as promised)
- Sinon (chai)

The library is conducive to TDD.  `grunt test:watch` is your friend. As modules (and templates) are exposed on their own namespaces you can easily isolate areas of the codebase for true unit testing without being forced to pull in the whole library or stub/mock modules irrelevent to the feature(s) you're testing.

#### Running Tests

##### Single run

```bash
grunt test
```

##### Watch

```bash
grunt test:watch
```

### Transpiling templates (html2js)

Transpiling our html templates into js allows us to neatly push them into the `$templateCache`.

To transpile the templates it's another simple grunt command:

```bash
grunt templates
```

This will compile the templates to the `dist/` folder. But it's probably best to avoid this all together. Both the `grunt test` and `grunt release` commands take care of all of this for you.

If you need to override the default template, simply replace what's already in the `$templateCache` with what ever you want. One way to achieve this is like this:

```html
<script id="templates/back-button.directive.tpl.html" type="text/html">
    <button>I'm a button!</button>
</script>
```

## Release Process

To make a release, ensure you have issued `grunt build`, committed the distribution package and tagged the commit with an appropriate version according to the [SemVer spec](http://semver.org/).

To make this easy for you, there's a handy grunt task. Simply issue `grunt release:major|minor|patch` and grunt will take care of building, committing and tagging for you. Then make a PR to the master branch of the upstream, merge upon CI build success and then all that's left to do is to push the tags to the upstream.

e.g:

```bash
  grunt release:minor
  git pull-request -b eHealthAfrica:master
  git push upstream --tags
```

### Publishing to npm

To publish a new version to npm, simply issue from the command line prior making a release (i.e.issuing a `grunt release` and pushing both commits and tags to the upstream):

```
npm publish
```

### Publishing to bower

Publishing to bower is slightly simpler in so far that you only have to do it once, and not explicitly for every release like npm:

e.g.

```
bower register angular-eha.image.directive https://github.com/eHealthAfrica/angular-eha.image.directivegit
```
## License

Copyright 2015 Matt Richards (eHealthAfrica)

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the License for the specific language governing permissions and limitations under the License.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - unreleased

## v1.0.3 - 19/03/2015

### Fixed
- Use ng-src instead of src to not throw a 404 request when first rendered

## v1.0.2 - 18/03/2015

### Fixed
- 'base64' instead of 'Base64'

## v1.0.1 - 03/03/2015

### Fixed
- Fix typo in package.json

## v1.0.0 - 03/03/2015

### Added
- Project boilerplate


tests/image.directive.spec.js
/*jshint expr: true*/
describe('Image directive', function() {
  var scope;
  var compile;
  var element;
  var imageData;

  beforeEach(module('eha.image.directive'));
  beforeEach(module('templates/image.directive.tpl.html'));
  beforeEach(inject(function(_$compile_, _$rootScope_, _$httpBackend_) {
    compile = _$compile_;
    scope = _$rootScope_.$new();
    scope.image = {
      mimeType: 'image/jpeg',
      data: '!!!PICTURE DATA!!!',
      title: 'The ALT text'
    };

    _$httpBackend_
    .expectGET('templates/image.directive.tpl.html')
    .respond();
  }));

  describe('should create dataURI for given object', function() {
    describe('image/jpeg', function() {
      beforeEach(function() {
        element = angular.element('<eha-img image="image" />');
        compile(element)(scope);
      });

      it('Passing an object by image attr.', function() {
        scope.$digest();
        var el = element.find('img')[0];
        expect(el.src).to.equal(
          encodeURI('data:image/jpeg;base64,!!!PICTURE DATA!!!')
        );
        expect(el.alt).to.equal(scope.image.title);
      });
    });

    describe('image/png', function() {
      beforeEach(function() {
        scope.image.mimeType = 'image/png',
        element = angular.element('<eha-img image="image" />');
        compile(element)(scope);
      });
      it('Passing an object by image attr.', function() {
        scope.$digest();
        var el = element.find('img')[0];
        expect(el.src).to.equal(
          encodeURI('data:image/png;base64,!!!PICTURE DATA!!!')
        );
      });
    });

    describe('unsafe', function() {
      // Check angular is escaping unsafe url schemes properly
      beforeEach(function() {
        scope.image.mimeType = 'text/png',
        element = angular.element('<eha-img image="image" />');
        compile(element)(scope);
      });
      it('Passing an object by image attr.', function() {
        scope.$digest();
        var el = element.find('img')[0];
        expect(el.src).to.equal(
          encodeURI('unsafe:data:text/png;base64,!!!PICTURE DATA!!!')
        );
      });
    });

  });

  describe('Passing alt', function() {
    describe('JSON object', function() {
      beforeEach(function() {
        element = angular.element('<eha-img image="image" alt="An image" />');
        compile(element)(scope);
      });

      it('should set the alt text', function() {
        scope.$digest();
        var el = element.find('img')[0];
        expect(el.alt).to.equal('An image');
      });

    });

    describe('Data URI', function() {
      beforeEach(function() {
        element = angular
                  .element('<eha-img src="http://example.com" alt="A URI" />');
        compile(element)(scope);
      });

      it('should set the alt text', function() {
        scope.$digest();
        var el = element.find('img')[0];
        expect(el.alt).to.equal('A URI');
      });
    });

  });

  describe('Passing a src', function() {
    beforeEach(function() {
      element = angular.element('<eha-img src="http://example.com" />');
      compile(element)(scope);
    });

    it('should render src in image tag', function() {
      scope.$digest();
      var el = element.find('img')[0];
      expect(el.src).to.equal('http://example.com/');
    });
  });
});


Gruntfile.js
module.exports = function(grunt) {

  require('load-grunt-tasks')(grunt);

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    clean: {
      dist: ['dist/'],
      tmp: ['.tmp/']
    },
    copy: {
      scripts: {
        expand: true,
        cwd: '.tmp',
        src: [
          'scripts.js',
          'scripts.templates.js'
        ],
        dest: 'dist/',
        rename: function(dest, src) {
          return dest + src.replace('scripts','angular-eha.image.directive');
        }
      }
    },
    concat: {
      scripts: {
        src: [
          'src/**/*.js',
          '!src/**/*.spec.js'
        ],
        dest: '.tmp/scripts.js'
      },
      templates: {
        src: [
          '.tmp/scripts.js',
          '.tmp/templates.js'
        ],
        dest: '.tmp/scripts.templates.js'
      }
    },
    ngAnnotate: {
      options: {
        singleQuotes: true
      },
      tmp: {
        files: [{
          expand: true,
          src: ['.tmp/**/*.js']
        }]
      }
    },
    uglify: {
      dist: {
        files: {
          'dist/angular-eha.image.directive.min.js': ['.tmp/scripts.js'],
          'dist/angular-eha.image.directive.templates.min.js': ['.tmp/scripts.templates.js']
        }
      }
    },
    html2js: {
      dist: {
        src: ['src/**/*.tpl.html'],
        dest: '.tmp/templates.js',
        module: 'eha.image.directive.templates',
        options: {
          rename: function(moduleName) {
            var parts = moduleName.split('/');
            var index = parts.indexOf('src');
            parts = parts.slice(index + 1, parts.length);
            return 'templates/' + parts.join('/');
          }
        }
      }
    },
    karma: {
      options: {
        configFile: 'karma.conf.js'
      },
      unit: {
        singleRun: true,
        autoWatch: false
      },
      watch: {
        singleRun: false,
        autoWatch: true
      }
    },
    bump: {
      options: {
        files: ['package.json', 'bower.json'],
        updateConfigs: [],
        commit: true,
        commitMessage: 'Release v%VERSION%',
        commitFiles: ['package.json', 'bower.json', 'dist/'],
        createTag: true,
        tagName: 'v%VERSION%',
        tagMessage: 'Version %VERSION%',
        push: false,
        // pushTo: 'origin',
        gitDescribeOptions: '--tags --always --abbrev=1 --dirty=-d',
        globalReplace: false,
        prereleaseName: false,
        regExp: false
      }
    },
    jshint: {
      all: ['src/**/*.js', 'tests/**/*.spec.js']
    },
    jscs: {
      src: ['src/**/*.js', 'tests/**/*.spec.js'],
      options: {
        config: ".jscsrc",
        requireCurlyBraces: [ "if" ]
      }
    }
  });

  grunt.registerTask('templates', ['html2js']);
  grunt.registerTask('test', ['html2js', 'jshint', 'jscs', 'karma:unit']);
  grunt.registerTask('test:watch', ['karma:watch']);

  grunt.registerTask('build', function() {
    grunt.task.run([
      'clean',
      'templates',
      'concat:scripts',
      'concat:templates',
      'ngAnnotate',
      'copy:scripts',
      'uglify:dist'
    ]);
  });

  grunt.registerTask('release', function(target) {
    grunt.task.run([
      'build',
      'bump:' + target
    ])
  })

  grunt.registerTask('default', ['jshint', 'jscs', 'build']);

};


src/image.directive.js
;(function() {
  'use strict';
  /**
   * @ngdoc directive
   * @restrict E
   * @name ehaImg
   * @module eha.image.directive
   * @element ehaImg
   * @description
   *
   * An `img` tag wrapper directive providing dataURI and URI support.
   *
   * Accepts either a JSON representation of an image (see:
   * https://github.com/eHealthAfrica/data_model/tree/master/schemas/Image.json)
   * passed via the `image` attribute or alternatively a URI passed via the
   * `src` attribute. `image.title` will be used as `img` alt text if passed
   * via `image` attribute.
   *
   * Also accepts an `alt` parameter that will be passed to the `img` tag
   * taking priority over `image.title` if both are set.
   *
   * @example
   * <example module="ehaImageExample1">
   *  <file name="demo1.js">
   *    angular.module('ehaImageExample', ['eha.image.directive', 'eha.image.directive.templates'])
   *    .controller('DemoCtrl1', function($scope) {
   *      $scope.image = {
   *        title: 'Test Image',
   *        mimeType: 'image/jpeg',
   *        type: 'jpeg',
   *        data: '/9j/4AAQSkZJRgABAQEASABIAAD/4QBARXhpZgAATU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAAqACAAQAAAABAAABGqADAAQAAAABAAABGgAAAAD/7QA4UGhvdG9zaG9wIDMuMAA4QklNBAQAAAAAAAA4QklNBCUAAAAAABDUHYzZjwCyBOmACZjs+EJ+/8AAEQgBGgEaAwEiAAIRAQMRAf/EAB8AAAEFAQEBAQEBAAAAAAAAAAABAgMEBQYHCAkKC//EALUQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+v/EAB8BAAMBAQEBAQEBAQEAAAAAAAABAgMEBQYHCAkKC//EALURAAIBAgQEAwQHBQQEAAECdwABAgMRBAUhMQYSQVEHYXETIjKBCBRCkaGxwQkjM1LwFWJy0QoWJDThJfEXGBkaJicoKSo1Njc4OTpDREVGR0hJSlNUVVZXWFlaY2RlZmdoaWpzdHV2d3h5eoKDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uLj5OXm5+jp6vLz9PX29/j5+v/bAEMAAQEBAQEBAgEBAgMCAgIDBAMDAwMEBQQEBAQEBQYFBQUFBQUGBgYGBgYGBgcHBwcHBwgICAgICQkJCQkJCQkJCf/bAEMBAQEBAgICBAICBAkGBQYJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCQkJCf/dAAQAEv/aAAwDAQACEQMRAD8A++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Q++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//R++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//S++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//T++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//U++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//V++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//W++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//X++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Q++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//R++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//S++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//T++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//U++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//V++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//W++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//X++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Q++KKKK/ynP8Ah3CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA//Z'
   *      };
   *    });
   *  </file>
   *  <file name="demo1.html">
   *
   *    <div ng-controller="DemoCtrl1">
   *      <eha-img image="image">
   *    </div>
   *  </file>
   * </example>
   *
   *
   * @example
   * <example module="ehaImageExample2">
   *  <file name="demo2.js">
   *    angular.module('ehaImageExample2', ['eha.common.directive.image', 'eha.common.templates'])
   *  </file>
   *  <file name="demo2.html">
   *    <eha-img src="http://placekitten.com/200/300">
   *  </file>
   * </example>
   */
  var ngModule = angular.module('eha.image.directive', [])
  .directive('ehaImg', function() {
    return {
      restrict: 'E',
      scope: {
        image: '=image',
        src: '@src',
        alt: '@alt'
      },
      templateUrl: 'templates/image.directive.tpl.html',
      link: function(scope) {
        if (scope.image) {
          if (scope.image.data) {
          // We've got data
            scope.src = encodeURI('data:' +
                                  scope.image.mimeType + ';' +
                                  'base64,' +
                                  scope.image.data);
          }

          if (scope.image.title && scope.alt === undefined) {
            // If alt text hasn't been passed via attribute, check the image
            // for a title and use that if available
            scope.alt = scope.image.title;
          }
        }
      }
    };
  });

  // Check for and export to commonjs environment
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = ngModule;
  }

}());
